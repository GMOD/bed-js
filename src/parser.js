import parser from "./autoSql";
import types from "./defaultTypes";
import { detectTypes } from "./util";

const strandMap = { ".": 0, "-": -1, "+": 1 };

export default class BED {
  constructor(args = {}) {
    if (args.autoSql) {
      this.autoSql = detectTypes(parser.parse(args.autoSql));
    } else if (args.type) {
      if (!types[args.type]) {
        throw new Error("Type not found");
      }
      this.autoSql = detectTypes(types[args.type]);
    } else {
      this.autoSql = detectTypes(types.defaultBedSchema);
      this.attemptDefaultBed = true;
    }
  }

  /*
   * parses a line of text as a BED line with the loaded autoSql schema
   *
   * @param line - a BED line as tab delimited text or array
   * @param opts - supply opts.uniqueId
   * @return a object representing a feature
   */
  parseLine(line, opts = {}) {
    const { autoSql } = this;
    const { uniqueId } = opts;
    let fields = line;
    if (!Array.isArray(line)) {
      if (line.startsWith("track") || line.startsWith("browser")) {
        throw new Error(
          `track and browser line parsing is not supported, please filter:\n${line}`
        );
      }
      fields = line.split("\t");
    }

    let feature = {};

    if (
      (this.attemptDefaultBed && fields.length === 12) ||
      !this.attemptDefaultBed
    ) {
      for (let i = 0; i < autoSql.fields.length; i++) {
        const autoField = autoSql.fields[i];
        let columnVal = fields[i];
        const { isNumeric, isArray, arrayIsNumeric, name } = autoField;
        if (columnVal === null || columnVal === undefined) {
          break;
        }
        if (columnVal !== ".") {
          if (isNumeric) {
            const num = Number(columnVal);
            columnVal = Number.isNaN(num) ? columnVal : num;
          } else if (isArray) {
            columnVal = columnVal.split(",");
            if (columnVal[columnVal.length - 1] === "") {
              columnVal.pop();
            }
            if (arrayIsNumeric) {
              columnVal = columnVal.map((str) => Number(str));
            }
          }

          feature[name] = columnVal;
        }
      }
    } else {
      const fieldNames = ["chrom", "chromStart", "chromEnd", "name"];
      feature = Object.fromEntries(
        fields.map((f, i) => [fieldNames[i] || "field" + i, f])
      );
      feature.chromStart = +feature.chromStart;
      feature.chromEnd = +feature.chromEnd;
      if (!Number.isNaN(Number.parseFloat(feature.field4))) {
        feature.score = +feature.field4;
        delete feature.field4;
      }
      if (feature.field5 === "+" || feature.field5 === "-") {
        feature.strand = feature.field5;
        delete feature.field5;
      }
    }
    if (uniqueId) {
      feature.uniqueId = uniqueId;
    }
    feature.strand = strandMap[feature.strand] || 0;

    feature.chrom = decodeURIComponent(feature.chrom);
    return feature;
  }
}
