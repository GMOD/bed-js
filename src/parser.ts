// @ts-expect-error
import parser from './autoSql.js'
import types from './defaultTypes.ts'
import { detectTypes, AutoSqlSchema, AutoSqlPreSchema } from './util.ts'

const strandMap = { '.': 0, '-': -1, '+': 1 }

// heuristic that a BED file is BED12 like...the number in col 10 is
// blockCount-like
function isBed12Like(fields: string[]) {
  return (
    fields.length >= 12 &&
    !Number.isNaN(Number.parseInt(fields[9], 10)) &&
    fields[10]?.split(',').filter(f => !!f).length ===
      Number.parseInt(fields[9], 10)
  )
}
export default class BED {
  public autoSql: AutoSqlSchema

  private attemptDefaultBed?: boolean

  constructor(arguments_: { autoSql?: string; type?: string } = {}) {
    if (arguments_.autoSql) {
      this.autoSql = detectTypes(
        parser.parse(arguments_.autoSql) as AutoSqlPreSchema,
      )
    } else if (arguments_.type) {
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      if (!types[arguments_.type]) {
        throw new Error('Type not found')
      }
      this.autoSql = detectTypes(types[arguments_.type])
    } else {
      this.autoSql = detectTypes(types.defaultBedSchema)
      this.attemptDefaultBed = true
    }
  }

  /*
   * parses a line of text as a BED line with the loaded autoSql schema
   *
   * @param line - a BED line as tab delimited text or array
   * @param opts - supply opts.uniqueId
   * @return a object representing a feature
   */
  parseLine(line: string | string[], options: { uniqueId?: string } = {}) {
    const { autoSql } = this
    const { uniqueId } = options
    const fields = Array.isArray(line) ? line : line.split('\t')

    let feature = {} as Record<string, any>
    if (
      !this.attemptDefaultBed ||
      // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
      (this.attemptDefaultBed && isBed12Like(fields))
    ) {
      for (let index = 0; index < autoSql.fields.length; index++) {
        const autoField = autoSql.fields[index]
        let columnValue: any = fields[index]
        const { isNumeric, isArray, arrayIsNumeric, name } = autoField
        if (columnValue === null || columnValue === undefined) {
          break
        }
        if (columnValue !== '.') {
          if (isNumeric) {
            const number_ = Number(columnValue)
            columnValue = Number.isNaN(number_) ? columnValue : number_
          } else if (isArray) {
            columnValue = columnValue.split(',')
            if (columnValue.at(-1) === '') {
              columnValue.pop()
            }
            if (arrayIsNumeric) {
              columnValue = columnValue.map(Number)
            }
          }

          feature[name] = columnValue
        }
      }
    } else {
      const fieldNames = ['chrom', 'chromStart', 'chromEnd', 'name']
      feature = Object.fromEntries(
        fields.map((f, index) => [fieldNames[index] || 'field' + index, f]),
      )
      feature.chromStart = +feature.chromStart
      feature.chromEnd = +feature.chromEnd
      if (!Number.isNaN(Number.parseFloat(feature.field4))) {
        feature.score = +feature.field4
        delete feature.field4
      }
      if (feature.field5 === '+' || feature.field5 === '-') {
        feature.strand = feature.field5
        delete feature.field5
      }
    }
    if (uniqueId) {
      feature.uniqueId = uniqueId
    }
    feature.strand = strandMap[feature.strand as keyof typeof strandMap] || 0

    feature.chrom = decodeURIComponent(feature.chrom)
    return feature
  }
}
