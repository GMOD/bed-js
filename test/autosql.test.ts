import parser from '../src/autoSql'

describe('autoSql parser', () => {
  it('simple example', () => {
    const s1 = `table addressBook
		"A simple address book"
    (
    string name;  "Name - first or last or both, we don't care"
    lstring address;  "Street address"
    string city;  "City"
    uint zipCode;  "A zip code is always positive, so can be unsigned"
    char[2] state;  "Just store the abbreviation for the state"
    )`

    const res = parser.parse(s1)
    expect(res).toMatchSnapshot()
  })
  it('resolves symbolic', () => {
    const s1 = `table symbolCols
    "example of enum and set symbolic columns"
    (
    int id;                                          "unique id"
    enum(male, female) sex;                          "enumerated column"
    set(cProg,javaProg,pythonProg,awkProg) skills;   "set column"
    )`
    const res = parser.parse(s1)
    expect(res).toMatchSnapshot()
  })

  it('resolves multiple values', () => {
    const s1 = `
    simple point
    "A three dimensional point"
      (
      float x;  "Horizontal coordinate"
      float y;  "Vertical coordinate"
      float z;  "In/out of screen coordinate"
      )`

    const s2 = `simple color
    "A red/green/blue format color"
      (
      ubyte red;  "Red value 0-255"
      ubyte green; "Green value 0-255"
      ubyte blue;  "Blue value 0-255"
      )`

    const s3 = `object face
    "A face of a three dimensional solid"
      (
      simple color color;  "Color of this face"
      int pointCount;    "Number of points in this polygon"
      uint[pointCount] points;   "Indices of points that make up face in polyhedron point array"
      )`

    const s4 = `table polyhedron
    "A solid three dimensional object"
      (
      int faceCount;  "Number of faces"
      object face[faceCount] faces; "List of faces"
      int pointCount; "Number of points"
      simple point[pointCount] points; "Array of points"
      )`

    expect(parser.parse(s1)).toMatchSnapshot()
    expect(parser.parse(s2)).toMatchSnapshot()
    expect(parser.parse(s3)).toMatchSnapshot()
    expect(parser.parse(s4)).toMatchSnapshot()
  })

  it('real world', () => {
    const s1 = `table hg18KGchr7
    "UCSC Genes for chr7 with color plus GeneSymbol and SwissProtID"
    (
    string  chrom;              "Reference sequence chromosome or scaffold"
    uint    chromStart; "Start position of feature on chromosome"
    uint    chromEnd;   "End position of feature on chromosome"
    string  name;               "Name of gene"
    uint    score;              "Score"
    char[1] strand;             "+ or - for strand"
    uint    thickStart; "Coding region start"
    uint    thickEnd;   "Coding region end"
    uint        reserved;       "Green on + strand, Red on - strand"
    string  geneSymbol; "Gene Symbol"
    string  spID;               "SWISS-PROT protein Accession number"
    )`
    expect(parser.parse(s1)).toMatchSnapshot()
  })
  it('clinvar CNV table', () => {
    const s1 = `table clinVarBed
"Browser extensible data (12 fields) plus information about a ClinVar entry"
    (
    string chrom;        "Chromosome (or contig, scaffold, etc.)"
    uint   chromStart;   "Start position in chromosome"
    uint   chromEnd;     "End position in chromosome"
    string name;         "Name of item"
    uint   score;      "Score from 0-1000"
    char[1] strand;    "+ or -"
    uint thickStart;   "Start of where display should be thick (start codon)"
    uint thickEnd;     "End of where display should be thick (stop codon)"
    uint reserved;     "Used as itemRgb as of 2004-11-22"
    int blockCount;    "Number of blocks"
    int[blockCount] blockSizes; "Comma separated list of block sizes"
    int[blockCount] chromStarts; "Start positions relative to chromStart"
    lstring origName;         "ClinVar Variation Report"
    string clinSign;         "Clinical significance"
    string reviewStatus;   "Review Status"
    string type;         "Type of Variant"
    string geneId;         "Gene Symbol"
    string snpId;         "dbSNP ID"
    string nsvId;         "dbVar ID"
    lstring rcvAcc;         "ClinVar Allele Submission"
    string testedInGtr;         "Genetic Testing Registry"
    lstring phenotypeList;         "Phenotypes"
    lstring phenotype;         "Phenotype identifiers"
    string origin;         "Data origin"
    string assembly;         "Genome assembly"
    string cytogenetic;         "Cytogenetic status"
    lstring hgvsCod;         "Nucleotide HGVS"
    lstring hgvsProt;         "Protein HGVS"
    string numSubmit;         "Number of submitters"
    string lastEval;         "Last evaluation"
    string guidelines;         "Guidelines"
    lstring otherIds;         "Other identifiers e.g. OMIM IDs, etc."
    string _mouseOver;         "Mouse over text, not shown"
    )
   `
    expect(parser.parse(s1)).toMatchSnapshot()
  })

  it('pli', () => {
    const s1 = `table pliMetrics
"bed12+5 for displaying gnomAD haploinsufficiency prediction scores"
    (
    string chrom;      "Reference sequence chromosome or scaffold"
    uint   chromStart; "Start position in chromosome"
    uint   chromEnd;   "End position in chromosome"
    string name;       "ENST or ENSG Name"
    uint   score;      "pLI score between 0-1000, or  -1 for NA"
    char[1] strand;    "strand of transcript"
    uint thickStart;   "Start of where display is thick"
    uint thickEnd;     "End of where display should be thick"
    uint itemRgb;    "Color of item"
    int blockCount;   "Number of exons"
    int[blockCount] blockSizes;  "Size of each exon"
    int[blockCount] chromStarts; "0-based start position of each exon"
    string _mouseOver;  "Mouseover label"
    float _pli;         "pLI value for filters"
    string geneName;   "Gene symbol"
    string synonymous; "Synonymous metrics"
    string missense;   "Missense metrics"
    string pLoF;       "Predicted Loss of Function metrics"
    )`
    expect(parser.parse(s1)).toMatchSnapshot()
  })

  // seen in https://hgdownload.soe.ucsc.edu/gbdb/hg38/clinvarSubLolly/clinvarSubLolly.bb
  test('comment', () => {
    const table = `table clinsub
"Clinvar Submissions"
    (
    #bed 9
    string chrom;      "Chromosome (or contig, scaffold, etc.)"
    uint   chromStart; "Start position in chromosome"
    uint   chromEnd;   "End position in chromosome"
    string name;       "Name of item"
    uint   score;      "Score from 0-1000"
    char[1] strand;    "+ or -"
    uint thickStart;   "Start of where display should be thick (start codon)"
    uint thickEnd;     "End of where display should be thick (stop codon)"
    uint reserved;     "Used as itemRgb as of 2004-11-22"

    #extra fields
    uint lollySize;    "Size of lollipop"
    lstring changes;     "changes
    lstring variantIds;     "variantIds
    lstring subIds;     "subIds
    lstring _mouseOver;     "mouseOver"
    )`

    expect(parser.parse(table)).toMatchSnapshot()
  })
})
