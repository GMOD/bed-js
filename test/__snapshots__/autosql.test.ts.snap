// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`autoSql parser > clinvar CNV table 1`] = `
{
  "comment": "Browser extensible data (12 fields) plus information about a ClinVar entry",
  "fields": [
    {
      "comment": "Chromosome (or contig, scaffold, etc.)",
      "name": "chrom",
      "type": "string",
    },
    {
      "comment": "Start position in chromosome",
      "name": "chromStart",
      "type": "uint",
    },
    {
      "comment": "End position in chromosome",
      "name": "chromEnd",
      "type": "uint",
    },
    {
      "comment": "Name of item",
      "name": "name",
      "type": "string",
    },
    {
      "comment": "Score from 0-1000",
      "name": "score",
      "type": "uint",
    },
    {
      "comment": "+ or -",
      "name": "strand",
      "size": 1,
      "type": "char",
    },
    {
      "comment": "Start of where display should be thick (start codon)",
      "name": "thickStart",
      "type": "uint",
    },
    {
      "comment": "End of where display should be thick (stop codon)",
      "name": "thickEnd",
      "type": "uint",
    },
    {
      "comment": "Used as itemRgb as of 2004-11-22",
      "name": "reserved",
      "type": "uint",
    },
    {
      "comment": "Number of blocks",
      "name": "blockCount",
      "type": "int",
    },
    {
      "comment": "Comma separated list of block sizes",
      "name": "blockSizes",
      "size": "blockCount",
      "type": "int",
    },
    {
      "comment": "Start positions relative to chromStart",
      "name": "chromStarts",
      "size": "blockCount",
      "type": "int",
    },
    {
      "comment": "ClinVar Variation Report",
      "name": "origName",
      "type": "lstring",
    },
    {
      "comment": "Clinical significance",
      "name": "clinSign",
      "type": "string",
    },
    {
      "comment": "Review Status",
      "name": "reviewStatus",
      "type": "string",
    },
    {
      "comment": "Type of Variant",
      "name": "type",
      "type": "string",
    },
    {
      "comment": "Gene Symbol",
      "name": "geneId",
      "type": "string",
    },
    {
      "comment": "dbSNP ID",
      "name": "snpId",
      "type": "string",
    },
    {
      "comment": "dbVar ID",
      "name": "nsvId",
      "type": "string",
    },
    {
      "comment": "ClinVar Allele Submission",
      "name": "rcvAcc",
      "type": "lstring",
    },
    {
      "comment": "Genetic Testing Registry",
      "name": "testedInGtr",
      "type": "string",
    },
    {
      "comment": "Phenotypes",
      "name": "phenotypeList",
      "type": "lstring",
    },
    {
      "comment": "Phenotype identifiers",
      "name": "phenotype",
      "type": "lstring",
    },
    {
      "comment": "Data origin",
      "name": "origin",
      "type": "string",
    },
    {
      "comment": "Genome assembly",
      "name": "assembly",
      "type": "string",
    },
    {
      "comment": "Cytogenetic status",
      "name": "cytogenetic",
      "type": "string",
    },
    {
      "comment": "Nucleotide HGVS",
      "name": "hgvsCod",
      "type": "lstring",
    },
    {
      "comment": "Protein HGVS",
      "name": "hgvsProt",
      "type": "lstring",
    },
    {
      "comment": "Number of submitters",
      "name": "numSubmit",
      "type": "string",
    },
    {
      "comment": "Last evaluation",
      "name": "lastEval",
      "type": "string",
    },
    {
      "comment": "Guidelines",
      "name": "guidelines",
      "type": "string",
    },
    {
      "comment": "Other identifiers e.g. OMIM IDs, etc.",
      "name": "otherIds",
      "type": "lstring",
    },
    {
      "comment": "Mouse over text, not shown",
      "name": "_mouseOver",
      "type": "string",
    },
  ],
  "name": "clinVarBed",
  "type": "table",
}
`;

exports[`autoSql parser > comment 1`] = `
{
  "comment": "Clinvar Submissions",
  "fields": [
    {
      "comment": "Chromosome (or contig, scaffold, etc.)",
      "name": "chrom",
      "type": "string",
    },
    {
      "comment": "Start position in chromosome",
      "name": "chromStart",
      "type": "uint",
    },
    {
      "comment": "End position in chromosome",
      "name": "chromEnd",
      "type": "uint",
    },
    {
      "comment": "Name of item",
      "name": "name",
      "type": "string",
    },
    {
      "comment": "Score from 0-1000",
      "name": "score",
      "type": "uint",
    },
    {
      "comment": "+ or -",
      "name": "strand",
      "size": 1,
      "type": "char",
    },
    {
      "comment": "Start of where display should be thick (start codon)",
      "name": "thickStart",
      "type": "uint",
    },
    {
      "comment": "End of where display should be thick (stop codon)",
      "name": "thickEnd",
      "type": "uint",
    },
    {
      "comment": "Used as itemRgb as of 2004-11-22",
      "name": "reserved",
      "type": "uint",
    },
    [
      [],
      "#",
      "extra fields",
      [
        "
",
        " ",
        " ",
        " ",
        " ",
      ],
    ],
    {
      "comment": "Size of lollipop",
      "name": "lollySize",
      "type": "uint",
    },
    {
      "comment": "changes",
      "name": "changes",
      "type": "lstring",
    },
    {
      "comment": "variantIds",
      "name": "variantIds",
      "type": "lstring",
    },
    {
      "comment": "subIds",
      "name": "subIds",
      "type": "lstring",
    },
    {
      "comment": "mouseOver",
      "name": "_mouseOver",
      "type": "lstring",
    },
  ],
  "name": "clinsub",
  "type": "table",
}
`;

exports[`autoSql parser > pli 1`] = `
{
  "comment": "bed12+5 for displaying gnomAD haploinsufficiency prediction scores",
  "fields": [
    {
      "comment": "Reference sequence chromosome or scaffold",
      "name": "chrom",
      "type": "string",
    },
    {
      "comment": "Start position in chromosome",
      "name": "chromStart",
      "type": "uint",
    },
    {
      "comment": "End position in chromosome",
      "name": "chromEnd",
      "type": "uint",
    },
    {
      "comment": "ENST or ENSG Name",
      "name": "name",
      "type": "string",
    },
    {
      "comment": "pLI score between 0-1000, or  -1 for NA",
      "name": "score",
      "type": "uint",
    },
    {
      "comment": "strand of transcript",
      "name": "strand",
      "size": 1,
      "type": "char",
    },
    {
      "comment": "Start of where display is thick",
      "name": "thickStart",
      "type": "uint",
    },
    {
      "comment": "End of where display should be thick",
      "name": "thickEnd",
      "type": "uint",
    },
    {
      "comment": "Color of item",
      "name": "itemRgb",
      "type": "uint",
    },
    {
      "comment": "Number of exons",
      "name": "blockCount",
      "type": "int",
    },
    {
      "comment": "Size of each exon",
      "name": "blockSizes",
      "size": "blockCount",
      "type": "int",
    },
    {
      "comment": "0-based start position of each exon",
      "name": "chromStarts",
      "size": "blockCount",
      "type": "int",
    },
    {
      "comment": "Mouseover label",
      "name": "_mouseOver",
      "type": "string",
    },
    {
      "comment": "pLI value for filters",
      "name": "_pli",
      "type": "float",
    },
    {
      "comment": "Gene symbol",
      "name": "geneName",
      "type": "string",
    },
    {
      "comment": "Synonymous metrics",
      "name": "synonymous",
      "type": "string",
    },
    {
      "comment": "Missense metrics",
      "name": "missense",
      "type": "string",
    },
    {
      "comment": "Predicted Loss of Function metrics",
      "name": "pLoF",
      "type": "string",
    },
  ],
  "name": "pliMetrics",
  "type": "table",
}
`;

exports[`autoSql parser > real world 1`] = `
{
  "comment": "UCSC Genes for chr7 with color plus GeneSymbol and SwissProtID",
  "fields": [
    {
      "comment": "Reference sequence chromosome or scaffold",
      "name": "chrom",
      "type": "string",
    },
    {
      "comment": "Start position of feature on chromosome",
      "name": "chromStart",
      "type": "uint",
    },
    {
      "comment": "End position of feature on chromosome",
      "name": "chromEnd",
      "type": "uint",
    },
    {
      "comment": "Name of gene",
      "name": "name",
      "type": "string",
    },
    {
      "comment": "Score",
      "name": "score",
      "type": "uint",
    },
    {
      "comment": "+ or - for strand",
      "name": "strand",
      "size": 1,
      "type": "char",
    },
    {
      "comment": "Coding region start",
      "name": "thickStart",
      "type": "uint",
    },
    {
      "comment": "Coding region end",
      "name": "thickEnd",
      "type": "uint",
    },
    {
      "comment": "Green on + strand, Red on - strand",
      "name": "reserved",
      "type": "uint",
    },
    {
      "comment": "Gene Symbol",
      "name": "geneSymbol",
      "type": "string",
    },
    {
      "comment": "SWISS-PROT protein Accession number",
      "name": "spID",
      "type": "string",
    },
  ],
  "name": "hg18KGchr7",
  "type": "table",
}
`;

exports[`autoSql parser > resolves multiple values 1`] = `
{
  "comment": "A three dimensional point",
  "fields": [
    {
      "comment": "Horizontal coordinate",
      "name": "x",
      "type": "float",
    },
    {
      "comment": "Vertical coordinate",
      "name": "y",
      "type": "float",
    },
    {
      "comment": "In/out of screen coordinate",
      "name": "z",
      "type": "float",
    },
  ],
  "name": "point",
  "type": "simple",
}
`;

exports[`autoSql parser > resolves multiple values 2`] = `
{
  "comment": "A red/green/blue format color",
  "fields": [
    {
      "comment": "Red value 0-255",
      "name": "red",
      "type": "ubyte",
    },
    {
      "comment": "Green value 0-255",
      "name": "green",
      "type": "ubyte",
    },
    {
      "comment": "Blue value 0-255",
      "name": "blue",
      "type": "ubyte",
    },
  ],
  "name": "color",
  "type": "simple",
}
`;

exports[`autoSql parser > resolves multiple values 3`] = `
{
  "comment": "A face of a three dimensional solid",
  "fields": [
    {
      "comment": "Color of this face",
      "name": "color",
      "type": "simple color",
    },
    {
      "comment": "Number of points in this polygon",
      "name": "pointCount",
      "type": "int",
    },
    {
      "comment": "Indices of points that make up face in polyhedron point array",
      "name": "points",
      "size": "pointCount",
      "type": "uint",
    },
  ],
  "name": "face",
  "type": "object",
}
`;

exports[`autoSql parser > resolves multiple values 4`] = `
{
  "comment": "A solid three dimensional object",
  "fields": [
    {
      "comment": "Number of faces",
      "name": "faceCount",
      "type": "int",
    },
    {
      "comment": "List of faces",
      "name": "faces",
      "size": "faceCount",
      "type": "object face",
    },
    {
      "comment": "Number of points",
      "name": "pointCount",
      "type": "int",
    },
    {
      "comment": "Array of points",
      "name": "points",
      "size": "pointCount",
      "type": "simple point",
    },
  ],
  "name": "polyhedron",
  "type": "table",
}
`;

exports[`autoSql parser > resolves symbolic 1`] = `
{
  "comment": "example of enum and set symbolic columns",
  "fields": [
    {
      "comment": "unique id",
      "name": "id",
      "type": "int",
    },
    {
      "comment": "enumerated column",
      "name": "sex",
      "type": "enum",
      "vals": [
        "male",
        "female",
      ],
    },
    {
      "comment": "set column",
      "name": "skills",
      "type": "set",
      "vals": [
        "cProg",
        "javaProg",
        "pythonProg",
        "awkProg",
      ],
    },
  ],
  "name": "symbolCols",
  "type": "table",
}
`;

exports[`autoSql parser > simple example 1`] = `
{
  "comment": "A simple address book",
  "fields": [
    {
      "comment": "Name - first or last or both, we don't care",
      "name": "name",
      "type": "string",
    },
    {
      "comment": "Street address",
      "name": "address",
      "type": "lstring",
    },
    {
      "comment": "City",
      "name": "city",
      "type": "string",
    },
    {
      "comment": "A zip code is always positive, so can be unsigned",
      "name": "zipCode",
      "type": "uint",
    },
    {
      "comment": "Just store the abbreviation for the state",
      "name": "state",
      "size": 2,
      "type": "char",
    },
  ],
  "name": "addressBook",
  "type": "table",
}
`;
